nucleus/common/glassfish-api/src/main/java/org/glassfish/api/invocation/InvocationManagerImpl.java

```diff
@@ -68,347 +71,334 @@
+            if (allTypesHandler != null) {
+                allTypesHandler.afterPreInvoke(type, prev, invocation);
             }
-        }
-
-        if (setCIH != null) {
-            for (int i = 0; i < setCIH.size(); i++) {
-                setCIH.get(i).getComponentInvocationHandler().afterPreInvoke(invocationType, previousInvocation, invocation);
+            if (typeHandler != null) {
+                typeHandler.afterPreInvoke(type, prev, invocation);
             }
         }
-
     }
 
+    @Override
     public <T extends ComponentInvocation> void postInvoke(T invocation) throws InvocationException {
-
-        // Get this thread's ArrayList
-        InvocationArray<ComponentInvocation> invocationArray = frames.get();
+        InvocationFrames frames = framesByThread.get();
         if (invocation.getInvocationType() == SERVICE_STARTUP) {
-            invocationArray.setInvocationAttribute(UN_INITIALIZED);
+            frames.setState(UN_INITIALIZED);
             return;
         }
 
-        int beforeSize = invocationArray.size();
-        if (beforeSize == 0) {
+        Iterator<ComponentInvocation> iter = frames.descendingIterator();
+        if (!iter.hasNext()) {
             throw new InvocationException();
         }
 
-        ComponentInvocation prevInv = beforeSize > 1 ? invocationArray.get(beforeSize - 2) : null;
-        ComponentInvocation curInv = invocationArray.get(beforeSize - 1);
+        iter.next(); // the last is the current is "invocation"
+        ComponentInvocation prev = iter.hasNext() ? iter.next() : null;
 
+        ComponentInvocationType type = invocation.getInvocationType();
+        ComponentInvocationHandler typeHandler = typeHandlers.get(type);
         try {
-            ComponentInvocationType invType = invocation.getInvocationType();
-
-            if (invHandlers != null) {
-                for (ComponentInvocationHandler handler : invHandlers) {
-                    handler.beforePostInvoke(invType, prevInv, curInv);
-                }
+            if (allTypesHandler != null) {
+                allTypesHandler.beforePostInvoke(type, prev, invocation);
             }
-
-            List<RegisteredComponentInvocationHandler> setCIH = regCompInvHandlerMap.get(invType);
-            if (setCIH != null) {
-                for (int i = 0; i < setCIH.size(); i++) {
-                    setCIH.get(i).getComponentInvocationHandler().beforePostInvoke(invType, prevInv, curInv);
-                }
+            if (typeHandler != null) {
+                typeHandler.beforePostInvoke(type, prev, invocation);
             }
-
         } finally {
             // pop the stack
-            invocationArray.remove(beforeSize - 1);
+            frames.removeLast();
 
-            if (invHandlers != null) {
-                for (ComponentInvocationHandler handler : invHandlers) {
-                    handler.afterPostInvoke(invocation.getInvocationType(), prevInv, invocation);
```

Comment:
NB. You might have noticed that `invocation` is passed to `afterPostInvoke` instead of `curInv`. I believe this is just a inconsistency in the chosen way to express the logic. At this point `curInv` is (or really should be) `invocation`. To avoid any future confusion I unified it to use `invocation` everywhere as it is done in `preInvoke` and as I think it makes most sense as this is what is passed to the method as "current". It is unfortunate how the API was designed and that it leaves room for inconsistencies like calling `postInvoke` with a different value then `preInvoke`. `postInvoke` really should not have had an argument as the current is that pushed to the stack when invoking `preInvoke`. Or the API should have made clear that the argument is only passed to `postInvoke` to verify that the nesting is correct and that it indeed is the current top of the stack.

Reply:
I think it is worth a warning when this assumption doesn't hold and you actually have a good spot since you already jump over the invocation head.

This might disclose bugs we have around nested transactions or cleanups.